netty-handler-4.2.0 requires org.bouncycastle.jsse in OSGi

Attempting to load netty-handler in OSGi leads to a splat:

```
2025-04-04T12:10:37,579 | ERROR | RMI TCP Connection(4)-127.0.0.1 | TestProbe                        | 98 - PAXEXAM-PROBE-87e2d067-0d99-4ecc-a39e-4af212af1b13 - 0.0.0 | Exception executing feature test
org.apache.felix.resolver.reason.ReasonException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=odl-netty-4; type=karaf.feature; version="[14.0.9.SNAPSHOT,14.0.9.SNAPSHOT]"; filter:="(&(osgi.identity=odl-netty-4)(type=karaf.feature)(version>=14.0.9.SNAPSHOT)(version<=14.0.9.SNAPSHOT))" [caused by: Unable to resolve odl-netty-4/14.0.9.SNAPSHOT: missing requirement [odl-netty-4/14.0.9.SNAPSHOT] osgi.identity; osgi.identity=io.netty.handler; type=osgi.bundle; version="[4.2.0.Final,4.2.0.Final]"; resolution:=mandatory [caused by: Unable to resolve io.netty.handler/4.2.0.Final: missing requirement [io.netty.handler/4.2.0.Final] osgi.wiring.package; filter:="(osgi.wiring.package=org.bouncycastle.jsse)"]]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1341) ~[?:?]
        at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:433) ~[?:?]
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:420) ~[?:?]
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:374) ~[?:?]
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:256) ~[?:?]
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:399) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve odl-netty-4/14.0.9.SNAPSHOT: missing requirement [odl-netty-4/14.0.9.SNAPSHOT] osgi.identity; osgi.identity=io.netty.handler; type=osgi.bundle; version="[4.2.0.Final,4.2.0.Final]"; resolution:=mandatory [caused by: Unable to resolve io.netty.handler/4.2.0.Final: missing requirement [io.netty.handler/4.2.0.Final] osgi.wiring.package; filter:="(osgi.wiring.package=org.bouncycastle.jsse)"]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1341) ~[?:?]
        ... 12 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve io.netty.handler/4.2.0.Final: missing requirement [io.netty.handler/4.2.0.Final] osgi.wiring.package; filter:="(osgi.wiring.package=org.bouncycastle.jsse)"
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1341) ~[?:?]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1341) ~[?:?]
        ... 12 more
```

This is due to `maven-bundle-plugin` deriving this requirement from Class.forName() invocations.

--- a/handler/pom.xml
+++ b/handler/pom.xml
@@ -88,6 +88,11 @@
       <optional>true</optional>
     </dependency>
     <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bctls-jdk18on</artifactId>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
       <groupId>${conscrypt.groupId}</groupId>
       <artifactId>${conscrypt.artifactId}</artifactId>
       <classifier>${conscrypt.classifier}</classifier>

--- a/pom.xml
+++ b/pom.xml
@@ -864,7 +864,28 @@
       </dependency>
 
       <!--
-        Completely optional integration with Bouncy Castle TLS.
+        Completely optional integration of netty-handler with Bouncy Castle.
+
+        In particular, BouncyCastleAlpnSslUtils is using
+        org.bouncycastle.jsse.{BCApplicationProtocolSelector,BCSSLEngine}
+        via reflection, loading them via Class.forName(String).
+
+        This method of use leads to zero bytecode-level traces of this
+        dependency and works for plain Java without any further reference to
+        bctls.
+
+        maven-bundle-plugin recognizes the invocation of Class.forName() parses
+        its argument to recognize the call requires the org.bouncycastle.jsse
+        to succeed. It generates an Import-Package stanza for it, resulting in
+        a bundle which does not resolve without bctls.
+
+        This declaration, along with its corresponding dependency declaration
+        in netty-handler, provides the additional information about our use.
+        Its presence allows maven-bundle-plugin to observe
+        'org.bouncycastle.jsse' to live in an optional dependency, which is
+        enough for it to also add ';resolution=optional' and thus netty-handler
+        will resolve successfully even when bctls is not present, matching our
+        intent.
       -->
       <dependency>
         <groupId>org.bouncycastle</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -863,6 +863,17 @@
         <optional>true</optional>
       </dependency>
 
+      <!--
+        Completely optional integration with Bouncy Castle TLS.
+      -->
+      <dependency>
+        <groupId>org.bouncycastle</groupId>
+        <artifactId>bctls-jdk18on</artifactId>
+        <version>${bouncycastle.version}</version>
+        <scope>compile</scope>
+        <optional>true</optional>
+      </dependency>
+
       <dependency>
         <groupId>com.fasterxml</groupId>
         <artifactId>aalto-xml</artifactId>

